// namespaces/dep/m1test/test/deeply.nested.groups.test.js
var assert = require('assert-extras'),
    common = require('./common');

var count = 0;

var tests = {
    group1: {
        'group 1 one OK': function(done) {
            count++;
            assert.ok(true);
            setTimeout(done, fakeDelay);
        },
        'group 1 mutliple OKs': function(done) {
            count++;
            assert.ok(true);
            assert.ok(true);
            assert.ok(true);
            setTimeout(done, fakeDelay);
        }
    },
    group2: {
        'group 2 one OK': function(done) {
            count++;
            assert.ok(true);
            setTimeout(done, fakeDelay);
        },
        'group 2 mutliple OKs': function(done) {
            count++;
            assert.ok(true);
            assert.ok(true);
            assert.ok(true);
            setTimeout(done, fakeDelay);
        }
    },
    group3: {
        group3_1: {
            'group 3_1 one OK': function(done) {
                count++;
                assert.ok(true);
                setTimeout(done, fakeDelay);
            },
            'group 3_1 mutliple OKs': function(done) {
                count++;
                assert.ok(true);
                assert.ok(true);
                assert.ok(true);
                assert.ok(true);
                assert.ok(true);
                setTimeout(done, fakeDelay);
            }
        },
        group3_2: {
            group3_2_1: {
                group3_2_1_1: {
                    group3_2_1_1_1: {
                        'group 3_2_1_1_1_1 one OK': function(done) {
                            count++;
                            assert.ok(true);
                            setTimeout(done, fakeDelay);
                        },
                        'group 3_2_1_1_1_2 mutliple OKs': function(done) {
                            count++;
                            assert.ok(true);
                            assert.ok(true);
                            assert.ok(true);
                            setTimeout(done, fakeDelay);
                        },
                        'group 3_2_1_1_1_3 mutliple OKs': function(done) {
                            count++;
                            assert.ok(true);
                            assert.ok(true);
                            assert.ok(true);
                            assert.ok(count === 10, 'count should be incremented to 10');
                            setTimeout(done, fakeDelay);
                        }
                    }
                }
            },
            'verify count inside 3_2': function(done) {
                count++;
                assert.ok(count === 7, 'count should be incremented to 6');
                done();
            }
        }
    }
};

module.exports = tests;
