// namespaces/dep/m1test/test/groups.before.and.after.hook.test.js
var assert = require('assert-extras'),
    common = require('./common');

var count = 0;

var tests = {
    after: function(done) {
        count++;
        done();
    },
    before: function(done) {
        count--;
        done();
    },
    group1: {
        after: function(done) {
            count++;
            done();
        },
        before: function(done) {
            count--;
            done();
        },
        'group 1 one OK': function(done) {
            assert.ok(true);
            setTimeout(done, fakeDelay);
        },
        'group 1 mutliple OKs': function(done) {
            assert.ok(true);
            assert.ok(true);
            assert.ok(true);
            setTimeout(done, fakeDelay);
        },
        'verify count': function(done) {
            assert.ok(count === -2, 'count should be incremented to -2');
            setTimeout(done, fakeDelay);
        }
    },
    group2: {
        'group 2 one OK': function(done) {
            assert.ok(true);
            setTimeout(done, fakeDelay);
        },
        'group 2 mutliple OKs': function(done) {
            assert.ok(true);
            assert.ok(true);
            assert.ok(true);
            setTimeout(done, fakeDelay);
        },
        'verify count': function(done) {
            assert.ok(count === -1, 'count should be incremented to -1');
            setTimeout(done, fakeDelay);
        }
    },
    group3: {
        after: function(done) {
            count++;
            done();
        },
        before: function(done) {
            count--;
            done();
        },
        group3_1: {
            after: function(done) {
                count++;
                done();
            },
            before: function(done) {
                count--;
                done();
            },
            'group 3_1 one OK': function(done) {
                assert.ok(true);
                setTimeout(done, fakeDelay);
            },
            'group 3_1 mutliple OKs': function(done) {
                assert.ok(true);
                assert.ok(true);
                assert.ok(true);
                assert.ok(true);
                assert.ok(true);
                setTimeout(done, fakeDelay);
            },
            'verify count': function(done) {
                assert.ok(count === -3, 'count should be incremented to -3');
                setTimeout(done, fakeDelay);
            }
        },
    }
};

module.exports = tests;
